### PR TITLE
Check errno as early as possible

### PR DESCRIPTION
**Pull request by asimrp**
Previously, the result of select() system call and errno set by it was checked after performing several function calls, including checking for interrupts and checkForCancelFromQD.  That made it very likely for errno to change, losing the original value that was set by the
select().

This patch fixes it so that the errno is checked immediately after the system call.  This should address intermittent failures in CI with error message like this:
```
ERROR","58M01","interconnect error: select: Success"
```

### CODE REVIEW

**Comment by hlinnaka** [comment]
Is it ok to skip the checking for interrupts, checkForCancelFromQD, and the verbose logging, if EINTR is happens? Maybe it would be safer to save the errno in a local variable before checking the interrupts etc, but keep the order otherwise the same?

**Comment by zhangh43** [comment]
I have a commit in 5X to fix this errno override issue, but it's not in master.
The errno will be overrided by system call and `checkForCancelFromQD` would call `poll` inside.
As a result, this commit is to place `ereport` ahead of `checkForCancelFromQD`.
If we received interrupt signal, we still call `ML_CHECK_FOR_INTERRUPTS` after `select` in this PR.
```
commit 7a31b2c326434af1ef2d65b64859ed63d0e17603
Author: Hubert Zhang <hubertzhang@apache.org>
Date:   Tue Oct 29 11:42:18 2019 +0800

    Check select return value before calling checkForCancelFromQD

    checkForCancelFromQD will call poll inside, which will override
    the errno. This could lead that select error reports error with
    misleading message: "interconnect error: select: Success"

diff --git a/src/backend/cdb/motion/ic_tcp.c b/src/backend/cdb/motion/ic_tcp.c
index 3da7da4582..c808915002 100644
--- a/src/backend/cdb/motion/ic_tcp.c
+++ b/src/backend/cdb/motion/ic_tcp.c
@@ -1636,8 +1636,6 @@ SetupTCPInterconnect(EState *estate)
                ML_CHECK_FOR_INTERRUPTS(estate->interconnect_context->teardownActive);
                n = select(highsock + 1, (fd_set *)&rset, (fd_set *)&wset, (fd_set *)&eset, &timeout);
                ML_CHECK_FOR_INTERRUPTS(estate->interconnect_context->teardownActive);
-               if (Gp_role == GP_ROLE_DISPATCH)
-                       checkForCancelFromQD(estate->interconnect_context);

                elapsed_ms = gp_get_elapsed_ms(&startTime);

@@ -1679,6 +1677,9 @@ SetupTCPInterconnect(EState *estate)
                                                        errmsg("Interconnect error: %s: %m", "select")));
                }

+               if (Gp_role == GP_ROLE_DISPATCH)
+                       checkForCancelFromQD(estate->interconnect_context);
+
```

**Comment by asimrp** [comment]
Thanks for the reviews.  My impression was that select system call shouldn't get interrupted frequently, so it's ok it skip checking cancel request from dispatched.  The cancel requests will be checked in the next iteration.  However, skipping the verbose logging is not right, as Heikki pointed out.  Please see a followup commit to address these concerns.

Hubert, I considered your 5X patch but then some future changes may introduce new function calls that inadvertently modify errno, rendering the fix ineffective.  It's better to remember the fact that select system call was interrupted right after the system call returns.

File: src/backend/cdb/motion/ic_tcp.c [hunk 1]
```diff
@@ -1331,6 +1331,7 @@ SetupTCPInterconnect(EState *estate)
 		int			highsock = -1;
 		uint64		timeout_ms = 20 * 60 * 1000;
 		int			outgoing_fail_count = 0;
+		bool		select_errno;
```
Comment (hlinnaka) [inline src/backend/cdb/motion/ic_tcp.c:1334]: wrong datatype

Comment (asimrp) [reply src/backend/cdb/motion/ic_tcp.c:1334]: Ooops! Thank you for catching it.  Fixed.  Surprised that clang 11.0.3 did not report a warning.

**Pull request merged**